### ISSUE TITLE
Generic constructor invocations from within class may call wrong version

### ISSUE THREAD

**Issue by vsmenon** [dart-lang/dev_compiler#178]
The following code throws an incorrect type error in DDC:

```dart
class A<T> {
  T x;

  A(y) {
    x = y;
  }

  foo() {
    return new A("$x");
  }
}

void main() {
  var a = new A<int>(42);
  var b = a.foo();
  print(b.x);
}
```

The problem is in the generated JavaScript where the A in foo implicitly refers to A<String> not A<dynamic>:

```javascript
  let A$ = dart.generic(function(T) {
    class A extends core.Object {
      A(y) {
        this.x = null;
        this.x = dart.as(y, T);
      }
      foo() {
        return new A(`${this.x}`);
      }
    }
    return A;
  });
  let A = A$();
```

I.e., we're picking up the inner definition of A (A<T>) not the outer definition of A (A<dynamic>) from the let.  We can solve by making calls to constructors expanded explicitly within the class.

**Comment by vsmenon** [comment]
Note, this comes up often in List and Future methods.  Once this is patched in:

https://codereview.chromium.org/1140523002/

we'll hit the same bug on just:

```dart
void main() {
  var l = <int>[1, 2, 3];
  var s = l.join();
  print(s);
}
```

**Comment by vsmenon** [comment]
Fix for review: https://codereview.chromium.org/1135543003/
